File: frontend/src/types/learner.ts
export interface LearningGoal {
  type?: string | null;
  description?: string | null;
  target?: string | number | null;
}

export interface LearningPreferences {
  preferred_sequence: string[];
  content_formats: string[];
  preferred_difficulty?: string | null;
}

export interface LearnerProfilePayload {
  education_level?: string | null;
  subject?: string | null;
  learning_goal?: LearningGoal | null;
  deadline?: string | null;
  current_level?: string | null;
  known_concepts?: string[] | null;
  weak_concepts?: string[] | null;
  misconceptions?: string[] | null;
  minutes_per_day?: number | null;
  days_per_week?: number | null;
  available_periods?: string[] | null;
  learning_preferences?: LearningPreferences | null;
  confidence_scores?: Record<string, number>;
}

export interface LearnerProfile {
  id: string;
  user_id: string;
  education_level: string | null;
  subject: string | null;
  learning_goal: LearningGoal;
  deadline: string | null;
  current_level: string | null;
  known_concepts: string[];
  weak_concepts: string[];
  misconceptions: string[];
  minutes_per_day: number | null;
  days_per_week: number | null;
  available_periods: string[];
  learning_preferences: LearningPreferences;
  diagnostic_results: Array<Record<string, unknown>>;
  confidence_scores: Record<string, number>;
  missing_fields: string[];
  profile_version: number;
  created_at: string;
  updated_at: string;
}

export interface UnderstandingResponse {
  profile_patch: LearnerProfilePayload;
  evidence: Array<Record<string, unknown>>;
  missing_fields: string[];
  contradictions: Array<Record<string, unknown>>;
  clarification_question: string | null;
  diagnostic_required: boolean;
  profile: LearnerProfile;
}

export interface LearningEventPayload {
  topic_id: string;
  correct: boolean;
  difficulty: number;
  hint_used: boolean;
  attempt_count: number;
  source: string;
}

export interface Mastery {
  topic_id: string;
  mastery_score: number;
  confidence: number;
  repeated_errors: number;
  level: string;
  last_assessed_at: string | null;
}

export interface ConceptInput {
  id: string;
  name: string;
  description?: string | null;
  difficulty: number;
  estimated_minutes: number;
  prerequisites: string[];
}

export interface RoadmapPayload {
  title?: string | null;
  target_concept_ids: string[];
  concepts: ConceptInput[];
  required_mastery: number;
  start_date: string;
}

export interface RoadmapItem {
  concept_id: string;
  title: string;
  sequence: number;
  session_number: number;
  planned_date: string;
  estimated_minutes: number;
  activity_type: string;
}

export interface Roadmap {
  id: string;
  status: string;
  title: string;
  subject: string;
  deadline: string | null;
  total_estimated_minutes: number;
  profile_version: number;
  learning_gaps: Array<Record<string, unknown>>;
  skipped_concepts: Array<Record<string, unknown>>;
  items: RoadmapItem[];
}

File: backend/app/models/learner.py
from datetime import date, datetime
from uuid import UUID

from sqlalchemy import (
    JSON,
    CheckConstraint,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID as PostgreSQLUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class LearnerProfile(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "learner_profiles"
    __table_args__ = (
        CheckConstraint(
            "minutes_per_day >= 10 AND minutes_per_day <= 600",
            name="learner_minutes_per_day_range",
        ),
        CheckConstraint(
            "days_per_week >= 1 AND days_per_week <= 7",
            name="learner_days_per_week_range",
        ),
    )

    user_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        unique=True,
        index=True,
        nullable=False,
    )
    education_level: Mapped[str | None] = mapped_column(String(100))
    subject: Mapped[str | None] = mapped_column(String(150), index=True)
    learning_goal: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    deadline: Mapped[date | None] = mapped_column(Date)
    current_level: Mapped[str | None] = mapped_column(String(100))
    known_concepts: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    weak_concepts: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    misconceptions: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    minutes_per_day: Mapped[int | None] = mapped_column(Integer)
    days_per_week: Mapped[int | None] = mapped_column(Integer)
    available_periods: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    learning_preferences: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    diagnostic_results: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    confidence_scores: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    missing_fields: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    profile_version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)


class LearnerEvidence(Base, UUIDPrimaryKeyMixin):
    __tablename__ = "learner_evidence"

    learner_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("learner_profiles.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    evidence_type: Mapped[str] = mapped_column(String(50), nullable=False)
    source: Mapped[str] = mapped_column(String(100), nullable=False)
    topic_id: Mapped[str | None] = mapped_column(String(255), index=True)
    field_name: Mapped[str | None] = mapped_column(String(100))
    value_json: Mapped[dict] = mapped_column(JSON, nullable=False)
    confidence: Mapped[float] = mapped_column(Float, default=1.0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class LearnerTopicMastery(Base):
    __tablename__ = "learner_topic_mastery"
    __table_args__ = (
        CheckConstraint(
            "mastery_score >= 0 AND mastery_score <= 1",
            name="mastery_score_range",
        ),
        CheckConstraint(
            "confidence >= 0 AND confidence <= 1",
            name="mastery_confidence_range",
        ),
    )

    learner_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("learner_profiles.id", ondelete="CASCADE"),
        primary_key=True,
    )
    topic_id: Mapped[str] = mapped_column(String(255), primary_key=True)
    mastery_score: Mapped[float] = mapped_column(Float, nullable=False)
    confidence: Mapped[float] = mapped_column(Float, nullable=False)
    repeated_errors: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    last_assessed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class MasteryHistory(Base, UUIDPrimaryKeyMixin):
    """Nhật ký thay đổi mastery — append-only, phục vụ đánh giá hiệu quả khuyến nghị + báo cáo tiến bộ theo thời gian."""
    __tablename__ = "mastery_history"

    learner_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("learner_profiles.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    topic_id: Mapped[str] = mapped_column(String(255), index=True, nullable=False)
    old_score: Mapped[float] = mapped_column(Float, nullable=False)
    new_score: Mapped[float] = mapped_column(Float, nullable=False)
    delta: Mapped[float] = mapped_column(Float, nullable=False)
    source: Mapped[str] = mapped_column(String(100), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class Roadmap(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "roadmaps"

    learner_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("learner_profiles.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    subject: Mapped[str] = mapped_column(String(150), nullable=False)
    status: Mapped[str] = mapped_column(String(30), default="active", nullable=False)
    deadline: Mapped[date | None] = mapped_column(Date)
    total_estimated_minutes: Mapped[int] = mapped_column(Integer, nullable=False)
    profile_version: Mapped[int] = mapped_column(Integer, nullable=False)
    context_json: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)


class RoadmapItem(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "roadmap_items"
    __table_args__ = (
        UniqueConstraint(
            "roadmap_id", "sequence", name="uq_roadmap_items_sequence"
        ),
    )

    roadmap_id: Mapped[UUID] = mapped_column(
        PostgreSQLUUID(as_uuid=True),
        ForeignKey("roadmaps.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    concept_id: Mapped[str] = mapped_column(String(255), index=True, nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    sequence: Mapped[int] = mapped_column(Integer, nullable=False)
    session_number: Mapped[int] = mapped_column(Integer, nullable=False)
    planned_date: Mapped[date] = mapped_column(Date, nullable=False)
    estimated_minutes: Mapped[int] = mapped_column(Integer, nullable=False)
    activity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(String(30), default="pending", nullable=False)
